openmdao/core/tests/test_dyn_sizing.py

```diff
@@ -0,0 +1,717 @@
+
+    def setup(self):
+        for i in range(self.n_inputs):
+            self.add_input(f"x{i+1}", shape_by_conn=True, copy_shape=f"y{i+1}")
+            self.add_output(f"y{i+1}", shape_by_conn=True, copy_shape=f"x{i+1}")
+
+    def compute(self, inputs, outputs):
+        for i in range(self.n_inputs):
+            outputs[f"y{i+1}"] = 2*inputs[f"x{i+1}"]
+
+
+class DistribComp(om.ExplicitComponent):
+    # a distributed component with inputs and outputs that are not dynamically shaped
+    def __init__(self, global_size, n_inputs=2):
+        super(DistribComp, self).__init__()
+        self.n_inputs = n_inputs
+        self.global_size = global_size
+        self.options['distributed'] = True
+
+    def setup(self):
+        # evenly distribute the variable over the procs
+        ave, res = divmod(self.global_size, self.comm.size)
+        sizes = [ave + 1 if p < res else ave for p in range(self.comm.size)]
+
+        for i in range(self.n_inputs):
+            self.add_input(f"x{i+1}", val=np.ones(sizes[rank]))
+            self.add_output(f"y{i+1}", val=np.ones(sizes[rank]))
+
+    def compute(self, inputs, outputs):
+        for i in range(self.n_inputs):
+            outputs[f"y{i+1}"] = (self.comm.rank + 1)*inputs[f"x{i+1}"]
+
+
+class DynShapeGroupSeries(om.Group):
+    # strings together some number of components in series.
+    # component type is determined by comp_class
+    def __init__(self, n_comps, n_inputs, comp_class):
+        super(DynShapeGroupSeries, self).__init__()
+        self.n_comps = n_comps
+        self.n_inputs = n_inputs
+        self.comp_class = comp_class
+
+        for icmp in range(1, self.n_comps + 1):
+            self.add_subsystem(f"C{icmp}", self.comp_class(n_inputs=self.n_inputs))
+
+        for icmp in range(1, self.n_comps):
+            for i in range(1, self.n_inputs + 1):
+                self.connect(f"C{icmp}.y{i}", f"C{icmp+1}.x{i}")
+
+
+class DynShapeGroupConnectedInputs(om.Group):
+    # contains some number of components with all of their matching inputs connected.
+    # component type is determined by comp_class
+    def __init__(self, n_comps, n_inputs, comp_class):
+        super(DynShapeGroupConnectedInputs, self).__init__()
+        self.n_comps = n_comps
+        self.n_inputs = n_inputs
+        self.comp_class = comp_class
+
+        for icmp in range(1, self.n_comps + 1):
+            self.add_subsystem(f"C{icmp}", self.comp_class(n_inputs=self.n_inputs),
+                               promotes_inputs=['*'])
+
+
+class TestDynShapes(unittest.TestCase):
+    def test_baseline_series(self):
+        # this is just a sized source and unsized sink, and we put a DynShapeGroupSeries in between them
+        p = om.Problem()
+        indep = p.model.add_subsystem('indep', om.IndepVarComp('x1', val=np.ones((2,3))))
+        indep.add_output('x2', val=np.ones((4,2)))
+        p.model.add_subsystem('Gdyn', DynShapeGroupSeries(3, 2, DynShapeComp))
```

Comment:
Have you tested what happens if you accidentally set `shape_by_conn` to your own name?  I imagine you will just get a 'cannot resolve shape' exception, so probably fine.

Reply:
Yeah, doing a `copy_shape` using your own name will either result in a failure to resolve, or if that variable happens to be connected to other variables by some other means (shape_by_conn or a resolvable variable connects to it by its own shape_by_conn or copy_shape) then it'll just work anyway.  Your comment also reminded me that I didn't have a test for when copy_shape is set to a non-existent variable, so I added that.